File: apps/ios/ExpCore/Sources/AgentWorkflow.swift
import Foundation

// EXP-850 — the two latest-wins slots the steer wire gained beside the feed
// (`background_tasks`, `workflow`), and the working caption every client
// derives from the `turn` slot. Foundation only (the ExpCore rule), so
// ExpCoreTests drives all of it.

// MARK: - background_tasks (§2)

/// One background task the agent is running (a `&`-launched shell command, a
/// workflow, a spawned agent). `kind` is a contract `backgroundTaskKind`
/// value; anything this build does not know folds onto `other`.
public struct AgentBackgroundTask: Equatable, Sendable, Identifiable {
    public let id: String
    public let kind: String
    public let description: String
    /// The launching tool_use_id, once the CLI names one — the first list
    /// frame for a task carries none.
    public let toolId: String?

    public init(id: String, kind: String, description: String, toolId: String? = nil) {
        self.id = id
        self.kind = kind
        self.description = description
        self.toolId = toolId
    }

    /// A wire `backgroundTaskKind`, folded onto `other` for anything this
    /// build does not know (never fatal: a newer engine may name a new kind).
    public static func kind(_ raw: Any?) -> String {
        guard let value = raw as? String,
              DomainContract.backgroundTaskKindValues.contains(value) else { return "other" }
        return value
    }
}

// MARK: - workflow (§3)

/// Contract `workflowStatus`.
public enum AgentWorkflowStatus: String, Sendable, CaseIterable {
    case running
    case completed
    case failed
    case stopped

    /// A wire status, defaulting to `running` — a card whose status this build
    /// cannot read still draws its agents.
    public static func parse(_ raw: Any?) -> AgentWorkflowStatus {
        guard let value = raw as? String, let status = AgentWorkflowStatus(rawValue: value)
        else { return .running }
        return status
    }

    public var isTerminal: Bool { self != .running }
}

/// Contract `workflowAgentState`.
public enum AgentWorkflowAgentState: String, Sendable, CaseIterable {
    case queued
    case running
    case done
    case error

    public static func parse(_ raw: Any?) -> AgentWorkflowAgentState {
        guard let value = raw as? String, let state = AgentWorkflowAgentState(rawValue: value)
        else { return .queued }
        return state
    }

    /// `done` and `error` both count as finished (the caption's `done/total`).
    public var isFinished: Bool { self == .done || self == .error }
}

/// One phase of a workflow run.
public struct AgentWorkflowPhase: Equatable, Sendable, Identifiable {
    public let index: Int
    public let title: String

    public var id: Int { index }

    public init(index: Int, title: String) {
        self.index = index
        self.title = title
    }
}

/// One agent of a workflow run, latest-wins per `index`.
public struct AgentWorkflowAgent: Equatable, Sendable, Identifiable {
    public let index: Int
    public let label: String
    public let phaseIndex: Int?
    /// The subagent id its `subagent` edges publish under — how the card finds
    /// this agent's nested events.
    public let agentId: String?
    public let model: String?
    public let state: AgentWorkflowAgentState
    public let tokens: Int?
    public let toolCalls: Int?
    public let durationMs: Int?
    public let lastTool: String?
    public let lastToolSummary: String?
    public let resultPreview: String?
    public let error: String?

    public var id: Int { index }

    public init(
        index: Int,
        label: String,
        phaseIndex: Int? = nil,
        agentId: String? = nil,
        model: String? = nil,
        state: AgentWorkflowAgentState = .queued,
        tokens: Int? = nil,
        toolCalls: Int? = nil,
        durationMs: Int? = nil,
        lastTool: String? = nil,
        lastToolSummary: String? = nil,
        resultPreview: String? = nil,
        error: String? = nil
    ) {
        self.index = index
        self.label = label
        self.phaseIndex = phaseIndex
        self.agentId = agentId
        self.model = model
        self.state = state
        self.tokens = tokens
        self.toolCalls = toolCalls
        self.durationMs = durationMs
        self.lastTool = lastTool
        self.lastToolSummary = lastToolSummary
        self.resultPreview = resultPreview
        self.error = error
    }
}

/// How many agents of one phase sit in each state — the phase strip's counts.
public struct AgentWorkflowPhaseCounts: Equatable, Sendable {
    public let queued: Int
    public let running: Int
    public let done: Int
    public let error: Int

    public init(queued: Int = 0, running: Int = 0, done: Int = 0, error: Int = 0) {
        self.queued = queued
        self.running = running
        self.done = done
        self.error = error
    }

    public var total: Int { queued + running + done + error }
}

/// EXP-850 §3: a claude `Workflow` tool call's progress, published as a
/// latest-wins `workflow` event keyed by the call's own id — so the card
/// renders IN PLACE of that tool row, never as a second row.
public struct AgentWorkflow: Equatable, Sendable, Identifiable {
    /// The `Workflow` tool_use_id — the same id as its `tool` row.
    public let id: String
    public let name: String
    public let description: String?
    public let status: AgentWorkflowStatus
    public let phases: [AgentWorkflowPhase]
    public let agents: [AgentWorkflowAgent]
    /// The terminal notification's own summary, when it carried one.
    public let summary: String?

    public init(
        id: String,
        name: String,
        description: String? = nil,
        status: AgentWorkflowStatus = .running,
        phases: [AgentWorkflowPhase] = [],
        agents: [AgentWorkflowAgent] = [],
        summary: String? = nil
    ) {
        self.id = id
        self.name = name
        self.description = description
        self.status = status
        self.phases = phases
        self.agents = agents
        self.summary = summary
    }

    /// §7's caption for this card, through the ONE shared derivation.
    public var caption: String {
        WorkflowCaption.caption(
            name: name,
            status: status.rawValue,
            phases: phases.map { WorkflowCaptionPhase(index: $0.index, title: $0.title) },
            agents: agents.map {
                WorkflowCaptionAgent(
                    index: $0.index, state: $0.state.rawValue, phaseIndex: $0.phaseIndex
                )
            }
        )
    }

    /// The agents of one phase, in publish order.
    public func agents(inPhase index: Int) -> [AgentWorkflowAgent] {
        agents.filter { $0.phaseIndex == index }
    }

    /// The phase strip's per-state counts.
    public func counts(inPhase index: Int) -> AgentWorkflowPhaseCounts {
        let rows = agents(inPhase: index)
        return AgentWorkflowPhaseCounts(
            queued: rows.filter { $0.state == .queued }.count,
            running: rows.filter { $0.state == .running }.count,
            done: rows.filter { $0.state == .done }.count,
            error: rows.filter { $0.state == .error }.count
        )
    }

    /// Every agent id this card owns — the subagent runs it nests (and the
    /// tabs it must never offer).
    public var agentIds: Set<String> {
        Set(agents.compactMap(\.agentId))
    }
}

/// EXP-850 §1/§2: one line of the compact strip above the composer — a
/// background task, or a `wait` tool row that has not settled.
public struct AgentStripLine: Equatable, Sendable, Identifiable {
    public enum Kind: Equatable, Sendable {
        case backgroundTask
        case wait
    }

    public let id: String
    public let kind: Kind
    /// The line VERBATIM (`↻ {description}` / `Waiting on {detail}`), locked
    /// ×4 — nothing per client re-words it.
    public let text: String

    public init(id: String, kind: Kind, text: String) {
        self.id = id
        self.kind = kind
        self.text = text
    }
}

extension AgentFeed {
    // MARK: - Latest-wins workflow map (§3)

    /// Fold one `workflow` event into the card list: latest-wins per id, in
    /// FIRST-APPEARANCE order (desktop `SteerFeed::workflows`), so the cards
    /// never reshuffle under the reader.
    public static func applyWorkflow(
        _ current: [AgentWorkflow], workflow: AgentWorkflow
    ) -> [AgentWorkflow] {
        var next = current
        if let at = next.firstIndex(where: { $0.id == workflow.id }) {
            next[at] = workflow
        } else {
            next.append(workflow)
        }
        return next
    }

    /// The NEWEST still-running card — what the working caption and the
    /// synced `agent_caption` speak for (desktop `running_workflow`).
    public static func runningWorkflow(_ workflows: [AgentWorkflow]) -> AgentWorkflow? {
        workflows.last { $0.status == .running }
    }

    // MARK: - The bottom strip (§1/§2)

    /// EXP-850 §1: the contract `toolKind` a `TaskOutput`/`Monitor` call
    /// wears — the wait rows the strip reports. (The generated contract ships
    /// the LIST; this names the one value the strip keys on.)
    public static let toolKindWait = "wait"

    // A background-task line is the bare description; the repeat glyph is
    // drawn by the view (the same concept every client draws), never text.
    /// The lead of an open wait row's line. Byte-identical ×4.
    public static let waitingOnPrefix = "Waiting on "

    /// The strip above the composer: one line per background task, then one
    /// per UNSETTLED `wait` tool row, in feed order. Empty = no strip at all.
    ///
    /// A wait row whose publisher named no detail falls back to the tool's own
    /// name — `Waiting on ` alone says nothing.
    public static func stripLines(
        backgroundTasks: [AgentBackgroundTask], feed: [AgentFeedItem]
    ) -> [AgentStripLine] {
        var lines = backgroundTasks.map { task in
            AgentStripLine(
                id: "task:\(task.id)",
                kind: .backgroundTask,
                text: task.description
            )
        }
        var seen = Set(lines.map(\.text))
        for item in feed {
            guard case let .tool(id, name, detail, _, callId, kind, settled, _, _, _) = item,
                  kind == toolKindWait, !settled else { continue }
            let label = (detail?.isEmpty == false) ? detail! : name
            let text = "\(waitingOnPrefix)\(label)"
            guard !seen.contains(text) else { continue }
            seen.insert(text)
            lines.append(AgentStripLine(id: "wait:\(callId ?? "\(id)")", kind: .wait, text: text))
        }
        return lines
    }

    // MARK: - The workflow card's captions (§3)

    /// One phase chip's counts: `2 done · 1 running · 1 queued · 1 failed`.
    /// Zero segments are omitted and the order is fixed, so two phases always
    /// read the same way round.
    public static func workflowPhaseCaption(_ counts: AgentWorkflowPhaseCounts) -> String {
        var segments: [String] = []
        if counts.done > 0 { segments.append("\(counts.done) done") }
        if counts.running > 0 { segments.append("\(counts.running) running") }
        if counts.queued > 0 { segments.append("\(counts.queued) queued") }
        if counts.error > 0 { segments.append("\(counts.error) failed") }
        return segments.joined(separator: WorkflowCaption.separator)
    }

    /// One agent row's trailing telemetry: `12.4k tokens · 7 tools · 1m 05s`,
    /// with whatever the agent did not report left out. Nil = nothing to draw.
    public static func workflowAgentTelemetry(_ agent: AgentWorkflowAgent) -> String? {
        var segments: [String] = []
        if let tokens = agent.tokens, tokens > 0 {
            segments.append("\(workingTokens(tokens)) tokens")
        }
        if let calls = agent.toolCalls, calls > 0 {
            segments.append("\(calls) tool\(calls == 1 ? "" : "s")")
        }
        if let duration = agent.durationMs, duration > 0 {
            segments.append(workingDuration(ms: duration))
        }
        return segments.isEmpty ? nil : segments.joined(separator: WorkflowCaption.separator)
    }

    // MARK: - The working caption (§5)

    /// The bare label a run with no `turn.startedAt` shows — an older
    /// publisher sends no start stamp, and a verb needs one. Byte-identical ×4.
    public static let workingFallbackLabel = "Working…"

    /// The verb this turn wears: `verbs[startedAt % verbs.count]`, so every
    /// client picks the SAME word for the same turn and it never flickers.
    public static func workingVerb(startedAt: Int) -> String {
        let verbs = DomainContract.steerWorkingVerbs
        guard !verbs.isEmpty else { return "Working" }
        let index = abs(startedAt % verbs.count)
        return verbs[index]
    }

    /// `37s` / `2m 04s` / `1h 03m` — the ×4 duration format. A negative span
    /// (clock skew) reads as `0s`.
    public static func workingDuration(ms: Int) -> String {
        let seconds = max(0, ms / 1000)
        if seconds < 60 { return "\(seconds)s" }
        let minutes = seconds / 60
        if minutes < 60 { return "\(minutes)m \(pad2(seconds % 60))s" }
        return "\(minutes / 60)h \(pad2(minutes % 60))m"
    }

    /// `812` / `2.0k` / `1.2M` — the ×4 token format. Truncated, never
    /// rounded up, so a count can never read `1000.0k`.
    public static func workingTokens(_ tokens: Int) -> String {
        let count = max(0, tokens)
        if count < 1000 { return "\(count)" }
        if count < 1_000_000 { return "\(tenths(count, per: 100))k" }
        return "\(tenths(count, per: 100_000))M"
    }

    /// `{verb}… ({duration} · ↓ {tokens} tokens)` — the trailing working row's
    /// ONE caption (§5). While a workflow RUNS the head is its §7 caption
    /// instead of a verb, with the same suffix.
    ///
    /// No `startedAt` (an older publisher) falls back to `Working…` with no
    /// suffix at all; a known start with no token count keeps the duration and
    /// drops the ` · ↓ N tokens` half.
    public static func workingCaption(
        startedAt: Int?, tokens: Int?, now: Date, workflow: AgentWorkflow? = nil
    ) -> String {
        let head = workflow?.caption
        guard let startedAt else { return head ?? workingFallbackLabel }
        let lead = head ?? "\(workingVerb(startedAt: startedAt))…"
        let elapsed = Int(now.timeIntervalSince1970 * 1000) - startedAt
        var group = workingDuration(ms: elapsed)
        if let tokens, tokens > 0 {
            group += "\(WorkflowCaption.separator)↓ \(workingTokens(tokens)) tokens"
        }
        return "\(lead) (\(group))"
    }

    private static func pad2(_ value: Int) -> String {
        value < 10 ? "0\(value)" : "\(value)"
    }

    /// One decimal place, TRUNCATED: 2000/100 = 20 → `2.0`, 1_234_567/100_000
    /// = 12 → `1.2`.
    private static func tenths(_ value: Int, per: Int) -> String {
        let scaled = value / per
        return "\(scaled / 10).\(scaled % 10)"
    }
}

File: apps/android/app/src/main/java/com/exponential/app/domain/AgentWorkflow.kt
package com.exponential.app.domain

// EXP-850: the two latest-wins slots the engine added beside `config_state`,
// `usage`, `rate_limit` and `turn` — the machine's background tasks, and one
// card per workflow the agent is running — plus the working caption every
// client derives from the turn slot. Compose-free and unit tested on their own;
// mirrored ×4 (web `lib/agent-feed.ts`, iOS `AgentFeed`, desktop
// `steer::frames` + `steer::workflow`).

/** EXP-850 (S2): one thing the CLI is running in the background — a shell
 *  command, a nested workflow, an agent. Latest-wins as a WHOLE list: the
 *  engine republishes everything still running, and an empty list closes the
 *  strip. */
data class BackgroundTask(
    val id: String,
    /** A contract `backgroundTaskKind` value; anything else reads as
     *  [BACKGROUND_TASK_KIND_OTHER] so an older client never chokes. */
    val kind: String,
    val description: String,
    /** The launching tool_use_id, once a `task_started` names one. */
    val toolId: String? = null,
)

const val BACKGROUND_TASK_KIND_OTHER = "other"

/** EXP-850 (S3): one phase of a workflow — latest-wins per [index]. */
data class WorkflowPhase(val index: Int, val title: String)

/** EXP-850 (S3): one agent of a workflow — latest-wins per [index]. Every
 *  telemetry field is optional: the engine fills what the CLI reported. */
data class WorkflowAgent(
    val index: Int,
    val label: String,
    val phaseIndex: Int? = null,
    /** The subagent id its nested events are tagged with, once it starts. */
    val agentId: String? = null,
    val model: String? = null,
    /** A contract `workflowAgentState` value. */
    val state: String,
    val tokens: Long? = null,
    val toolCalls: Int? = null,
    val durationMs: Long? = null,
    val lastTool: String? = null,
    val lastToolSummary: String? = null,
    val resultPreview: String? = null,
    val error: String? = null,
) {
    /** Done or errored — the states [WorkflowCaption] counts as finished. */
    val isFinished: Boolean
        get() = state == WORKFLOW_AGENT_STATE_DONE || state == WORKFLOW_AGENT_STATE_ERROR
}

/** EXP-850 (S3): one workflow card. Latest-wins PER ID — the engine
 *  republishes the WHOLE state on every change, so a newer frame simply
 *  replaces the older one. */
data class WorkflowState(
    val id: String,
    val name: String,
    val description: String? = null,
    /** A contract `workflowStatus` value. */
    val status: String,
    val phases: List<WorkflowPhase> = emptyList(),
    val agents: List<WorkflowAgent> = emptyList(),
    val summary: String? = null,
) {
    val isRunning: Boolean get() = status == WORKFLOW_STATUS_RUNNING
}

const val WORKFLOW_STATUS_RUNNING = "running"
const val WORKFLOW_STATUS_COMPLETED = "completed"
const val WORKFLOW_STATUS_FAILED = "failed"
const val WORKFLOW_STATUS_STOPPED = "stopped"

const val WORKFLOW_AGENT_STATE_QUEUED = "queued"
const val WORKFLOW_AGENT_STATE_RUNNING = "running"
const val WORKFLOW_AGENT_STATE_DONE = "done"
const val WORKFLOW_AGENT_STATE_ERROR = "error"

/** How many workflow cards one session keeps — the journal's and the relay's
 *  cap (`JOURNAL_WORKFLOW_CAP` / `WORKFLOW_SLOT_CAP`), so a client never holds
 *  more than a replay could ever hand it. Oldest id evicted first. */
const val WORKFLOW_SLOT_CAP = 16

/** EXP-850 (S3): one phase of the card's phase strip — its title and how its
 *  agents are doing. The counts are what the strip renders; a phase with no
 *  agents still renders (the workflow declared it). */
data class WorkflowPhaseCounts(
    val index: Int,
    val title: String,
    val queued: Int = 0,
    val running: Int = 0,
    val done: Int = 0,
    val error: Int = 0,
) {
    val total: Int get() = queued + running + done + error
}

/**
 * The phase strip: every declared phase in index order with its agents counted
 * by state, plus — last — a synthetic "no phase" entry when agents name a phase
 * this workflow never declared (or name none at all), so no agent is ever
 * silently missing from the strip.
 */
fun workflowPhaseCounts(workflow: WorkflowState): List<WorkflowPhaseCounts> {
    val declared = workflow.phases.sortedBy { it.index }
    val known = declared.map { it.index }.toSet()
    fun counts(index: Int?, title: String): WorkflowPhaseCounts {
        val agents = workflow.agents.filter {
            if (index == null) it.phaseIndex == null || it.phaseIndex !in known else it.phaseIndex == index
        }
        return WorkflowPhaseCounts(
            index = index ?: -1,
            title = title,
            queued = agents.count { it.state == WORKFLOW_AGENT_STATE_QUEUED },
            running = agents.count { it.state == WORKFLOW_AGENT_STATE_RUNNING },
            done = agents.count { it.state == WORKFLOW_AGENT_STATE_DONE },
            error = agents.count { it.state == WORKFLOW_AGENT_STATE_ERROR },
        )
    }
    val rows = declared.map { counts(it.index, it.title) }
    val loose = counts(null, WORKFLOW_UNPHASED_TITLE)
    return if (loose.total > 0) rows + loose else rows
}

/** What the strip calls the agents that belong to no declared phase. */
const val WORKFLOW_UNPHASED_TITLE = "Other"

/** EXP-850 (S3): one phase's counts as the strip words them — done, running,
 *  queued, failed, in that order, omitting the zeroes. A declared phase with
 *  no agents yet says so rather than rendering an empty column. */
fun workflowPhaseSummary(counts: WorkflowPhaseCounts): String {
    val parts = listOfNotNull(
        counts.done.takeIf { it > 0 }?.let { "$it done" },
        counts.running.takeIf { it > 0 }?.let { "$it running" },
        counts.queued.takeIf { it > 0 }?.let { "$it queued" },
        counts.error.takeIf { it > 0 }?.let { "$it failed" },
    )
    return if (parts.isEmpty()) WORKFLOW_PHASE_EMPTY else parts.joinToString(WorkflowCaption.SEPARATOR)
}

/** A declared phase nothing has been queued into yet. */
const val WORKFLOW_PHASE_EMPTY = "no agents"

// ── EXP-850 (S5): the working caption ────────────────────────────────────────

/** The caption with nothing to say yet — no turn start, so no verb and no
 *  clock. Byte-identical ×4. */
const val WORKING_FALLBACK_CAPTION = "Working…"

/** The verb of the turn that started at [startedAt] — a stable pick out of the
 *  contract's 24, so the same turn reads the same on every client and never
 *  flickers between frames. Null without a start. */
fun workingVerb(startedAt: Long?): String? {
    if (startedAt == null) return null
    val verbs = DomainContract.steerWorkingVerbs
    if (verbs.isEmpty()) return null
    // `startedAt` is an epoch in ms, so this is a plain modulo; the abs keeps
    // a nonsense negative stamp on a real index (web parity).
    return verbs[Math.floorMod(Math.abs(startedAt), verbs.size.toLong()).toInt()]
}

/**
 * How long the turn has been going: `37s`, `2m 04s`, `1h 03m`. Absent input
 * renders nothing; a clock BEHIND the turn's start (skew) reads as `0s`.
 */
fun formatWorkingDuration(startedAt: Long?, nowMs: Long): String? {
    if (startedAt == null || startedAt <= 0L) return null
    return formatDurationMs(nowMs - startedAt)
}

/** The same clock over a plain elapsed span (a workflow agent's `durationMs`):
 *  `37s`, `2m 04s`, `1h 03m`. Sub-second and negative spans read as `0s`
 *  rather than disappearing (web `formatTurnDuration` parity — clock skew is
 *  not a reason for the caption to lose half its words). */
fun formatDurationMs(ms: Long?): String? {
    if (ms == null) return null
    val seconds = maxOf(0L, ms / 1000L)
    return when {
        seconds < 60L -> "${seconds}s"
        seconds < 3600L -> "${seconds / 60L}m ${(seconds % 60L).toString().padStart(2, '0')}s"
        else -> "${seconds / 3600L}h ${((seconds % 3600L) / 60L).toString().padStart(2, '0')}m"
    }
}

/** EXP-850 (S3): what one workflow agent row says about itself under its
 *  label — its model, its tokens, its tool calls and how long it took, in that
 *  order, omitting whatever the engine did not report. Null when it reported
 *  nothing at all. */
fun workflowAgentMetrics(agent: WorkflowAgent): String? {
    val parts = listOfNotNull(
        agent.model?.takeIf { it.isNotBlank() },
        formatWorkingTokens(agent.tokens)?.let { "$it tokens" },
        agent.toolCalls?.takeIf { it > 0 }
            ?.let { "$it tool call${if (it == 1) "" else "s"}" },
        formatDurationMs(agent.durationMs?.takeIf { it > 0L }),
    )
    return parts.takeIf { it.isNotEmpty() }?.joinToString(WorkflowCaption.SEPARATOR)
}

/** EXP-850 (S3): the one line an agent row shows about what it is doing (or
 *  did) — the failure, the result, or the tool it is on. */
fun workflowAgentNote(agent: WorkflowAgent): String? = when (agent.state) {
    WORKFLOW_AGENT_STATE_ERROR -> agent.error ?: agent.resultPreview
    WORKFLOW_AGENT_STATE_DONE -> agent.resultPreview
    WORKFLOW_AGENT_STATE_RUNNING -> agent.lastToolSummary ?: agent.lastTool
    else -> null
}?.takeIf { it.isNotBlank() }

/**
 * The turn's output tokens: `812`, `2.0k`, `1.2M`. TRUNCATED at one decimal
 * rather than rounded, so a count never reads as the next unit before it has
 * reached it (999_999 is `999.9k`, not `1000.0k`).
 */
fun formatWorkingTokens(tokens: Long?): String? {
    if (tokens == null || tokens < 0L) return null
    return when {
        tokens < 1_000L -> tokens.toString()
        tokens < 1_000_000L -> "${tenths(tokens, 1_000L)}k"
        else -> "${tenths(tokens, 1_000_000L)}M"
    }
}

/** `value / unit` to one TRUNCATED decimal, e.g. 1234/1000 → `1.2`. */
private fun tenths(value: Long, unit: Long): String {
    val scaled = value * 10L / unit
    return "${scaled / 10L}.${scaled % 10L}"
}

/**
 * EXP-850 (S5): the ONE caption the working row renders —
 * `{verb}… ({duration} · ↓ {tokens} tokens)`.
 *
 * [workflowCaption] (a running workflow's [WorkflowCaption]) REPLACES the verb
 * when there is one: while a workflow runs, what the agent is doing is the
 * workflow, not a mood word. The bracketed group is omitted while both the
 * duration and the token count are unknown, and each half is dropped on its
 * own — a codex run publishes neither and degrades to plain "Working…".
 */
fun workingCaption(
    startedAt: Long?,
    tokens: Long?,
    nowMs: Long,
    workflowCaption: String? = null,
): String {
    val head = workflowCaption?.takeIf { it.isNotBlank() }
        ?: workingVerb(startedAt)?.let { "$it…" }
        ?: WORKING_FALLBACK_CAPTION
    val parts = listOfNotNull(
        formatWorkingDuration(startedAt, nowMs),
        formatWorkingTokens(tokens)?.let { "↓ $it tokens" },
    )
    if (parts.isEmpty()) return head
    return "$head (${parts.joinToString(WorkflowCaption.SEPARATOR)})"
}

/** EXP-850 (S1): what an unsettled `wait` tool row says in the strip above the
 *  composer. The row itself stays an ordinary tool row in the transcript. */
fun waitingLabel(detail: String): String = "Waiting on $detail"

/** EXP-850 (S2): what one background task says in that same strip. */
fun backgroundTaskLabel(description: String): String = "↻ $description"
